File: 프로그래머스/lv2/67257. ［카카오 인턴］ 수식 최대화/［카카오 인턴］ 수식 최대화.java
import java.util.*;

class Solution {
    String[] ex;
    
    public long solution(String expression) {
        StringTokenizer st = new StringTokenizer(expression, "[-+*/]", true);
        ex = new String[st.countTokens()];
        for(int i = 0; i < ex.length; i++) ex[i] = st.nextToken();
        
        long max = 0;
        String[] operations = { "+-*", "+*-", "-+*", "-*+", "*-+", "*+-" };
        for(String oper : operations) max = Math.max(max, calc(oper));
        return max;
    }
    
    long calc(String str){
        String[] arr = str.split("");
        Deque<String> dq = new LinkedList<>();
        for(String s : ex) dq.addLast(s);
        for(int i = 0; i < 3; i++){
            Deque<String> tmp = new LinkedList<>();
            while(!dq.isEmpty()){
                String poll = dq.pollFirst();
                if(poll.equals(arr[i])){
                    String a = tmp.pollLast(), b = dq.pollFirst();
                    long rst = 0;
                    switch(arr[i]){
                        case "+" : rst = Long.parseLong(a) + Long.parseLong(b); break;
                        case "*" : rst = Long.parseLong(a) * Long.parseLong(b); break;
                        case "-" : rst = Long.parseLong(a) - Long.parseLong(b); break;
                    }
                    tmp.addLast(rst + "");
                } else{
                    tmp.addLast(poll);
                }
            }
            while(!tmp.isEmpty()) dq.addLast(tmp.pollFirst());
        }
        return Math.abs(Long.parseLong(dq.poll()));
    }
}

File: 프로그래머스/lv2/67257. ［카카오 인턴］ 수식 최대화/［카카오 인턴］ 수식 최대화.js
function solution(expression, max = 0) {
    const priority = [ '+-*', '+*-', '-+*', '-*+', '*-+', '*+-'];
    const tokens = expression.match(/([0-9]+|[\+\-\*])/g);
    for(const p of priority) {
        let sub = [ ...tokens ];
        for(const oper of [ ...p ]) {
            let idx = -1;
            while((idx = sub.indexOf(oper)) > 0) {
                sub = [
                    ...sub.slice(0, idx - 1),
                    eval(sub.slice(idx - 1, idx + 2).join('')),
                    ...sub.slice(idx + 2),
                ];
            }
        }
        max = Math.max(max, Math.abs(Number(sub[0])));
    }
    return max;
}